sdk/src/azure/core/az_mqtt5_rpc_client_codec.c

```diff
@@ -1,83 +1,420 @@
+                ? az_span_size(az_mqtt5_rpc_any_executor_id)
+                : az_span_size(executor_id));
+    required_length -= az_span_size(az_mqtt5_rpc_executor_id_key);
+  }
+  if (command_id_index != -1)
+  {
+    required_length
+        += (az_span_is_content_equal(command_name, AZ_SPAN_EMPTY)
+                ? az_span_size(client->_internal.command_name)
+                : az_span_size(command_name));
+    required_length -= az_span_size(az_mqtt5_rpc_command_id_key);
+  }
+
+  _az_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_topic_span, required_length);
+
+  az_span remainder = mqtt_topic_span;
+
+  for (int32_t i = 0; i < az_span_size(topic_format); i++)
+  {
+    uint8_t c = az_span_ptr(topic_format)[i];
+
+    if (c == '{')
+    {
+      if (i == invoker_client_index)
+      {
+        remainder = az_span_copy(remainder, client->_internal.client_id);
+        i += az_span_size(az_mqtt5_rpc_client_id_key) - 1;
+      }
+      else if (i == model_id_index)
+      {
+        remainder = az_span_copy(remainder, client->_internal.model_id);
+        i += az_span_size(az_mqtt5_rpc_service_id_key) - 1;
+      }
+      else if (i == executor_id_index)
+      {
+        if (az_span_is_content_equal(executor_id, AZ_SPAN_EMPTY))
+        {
+          remainder = az_span_copy(remainder, az_mqtt5_rpc_any_executor_id);
+          i += az_span_size(az_mqtt5_rpc_executor_id_key) - 1;
+        }
+        else
+        {
+          remainder = az_span_copy(remainder, executor_id);
+          i += az_span_size(az_mqtt5_rpc_executor_id_key) - 1;
+        }
+      }
+      else if (i == command_id_index)
+      {
+        remainder = (az_span_is_content_equal(command_name, AZ_SPAN_EMPTY))
+            ? az_span_copy(remainder, client->_internal.command_name)
+            : az_span_copy(remainder, command_name);
+        i += az_span_size(az_mqtt5_rpc_command_id_key) - 1;
+      }
+    }
+    else
+    {
+      remainder = az_span_copy_u8(remainder, c);
+    }
+  }
+
+  az_span_copy_u8(remainder, '\0');
+
+  if (out_mqtt_topic_length)
+  {
+    *out_mqtt_topic_length = (size_t)required_length;
+  }
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_mqtt5_rpc_client_codec_get_subscribe_topic(
+    az_mqtt5_rpc_client_codec* client,
+    char* mqtt_topic,
+    size_t mqtt_topic_size,
+    size_t* out_mqtt_topic_length)
 {
-  return az_rpc_get_topic_from_format(
-      client->_internal.options.subscription_topic_format,
-      client->_internal.model_id,
-      server_client_id,
-      client->_internal.client_id,
-      _az_span_is_valid(client->_internal.command_name, 1, 0) ? client->_internal.command_name
-                                                              : command_name,
-      out_response_topic,
-      out_topic_length);
+  _az_PRECONDITION_NOT_NULL(client);
+  _az_PRECONDITION_NOT_NULL(mqtt_topic);
+  _az_PRECONDITION(mqtt_topic_size > 0);
+
+  az_span topic_format = client->_internal.options.subscription_topic_format;
```

Comment:
Nit: this is very inefficient as it scans the strings several times.
A single-scan state machine would be advised, one that can be reused by both of the `get_*_topic` functions given that the tokens are always the same.

Reply:
Response topic and publish topic could be unique per invocation (although those two might be able to be created together?), and are generated at a different time than the subscribe topic, which would only be generated once

Reply:
Thanks for the suggestions! I will improve the parser in a future PR